transports/quic/src/connection.rs

```diff
@@ -0,0 +1,310 @@
+macro_rules! span {
+    ($name:expr, side = $e:expr) => {
+        let span = tracing::trace_span!($name, side = $e);
+        let _guard = span.enter();
+    };
+    ($name:expr, $inner:expr, $id:expr) => {
+        let span = tracing::trace_span!($name, side = debug($inner.side()), id = debug(&$id.id));
+        let _guard = span.enter();
+    };
+}
+#[cfg(not(feature = "tracing"))]
+macro_rules! span {
+    ($name:expr, $($id:ident = $e:expr),*) => {};
+    ($name:expr, $inner:expr, $id:expr) => {};
+}
+
+/// An outbound QUIC substream. This will eventually resolve to either a
+/// [`Substream`] or an [`Error`].
+#[derive(Debug)]
+pub struct Outbound(OutboundInner);
+
+impl StreamMuxer for QuicMuxer {
+    type OutboundSubstream = Outbound;
+    type Substream = Substream;
+    type Error = Error;
+    fn open_outbound(&self) -> Self::OutboundSubstream {
+        let mut inner = self.inner();
+        Outbound(if let Err(e) = inner.close_reason() {
+            OutboundInner::Complete(Err(e))
+        } else {
+            match inner.get_pending_stream() {
+                Either::Left(id) => OutboundInner::Complete(Ok(id)),
+                Either::Right(receiver) => OutboundInner::Pending(receiver),
+            }
+        })
+    }
+
+    fn destroy_outbound(&self, outbound: Outbound) {
+        let mut inner = self.inner();
+        let id = match outbound.0 {
+            OutboundInner::Complete(Err(_)) => return,
+            OutboundInner::Complete(Ok(id)) => id,
+            // try_recv is race-free, because the lock on `self` prevents
+            // other tasks from sending on the other end of the channel.
+            OutboundInner::Pending(mut channel) => match channel.try_recv() {
+                Ok(Some(id)) => id,
+                Err(oneshot::Canceled) | Ok(None) => return,
+            },
+            OutboundInner::Done => return,
+        };
+        inner.destroy_stream(id)
+    }
+
+    fn destroy_substream(&self, substream: Self::Substream) {
+        let mut inner = self.inner();
+        span!("destroy", inner, substream);
+        inner.destroy_stream(substream.id)
+    }
+
+    fn is_remote_acknowledged(&self) -> bool {
+        // we do not allow 0RTT traffic, for security reasons, so this is
+        // always true.
+        true
+    }
+
+    fn poll_inbound(&self, cx: &mut Context<'_>) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner();
+        span!("inbound", side = debug(inner.side()));
+        trace!("being polled for inbound connections!");
+        inner.close_reason()?;
+        inner.wake_driver();
+        let stream = ready!(inner.accept(cx));
+        Poll::Ready(Ok(Substream::live(stream)))
+    }
+
+    fn write_substream(
+        &self,
+        cx: &mut Context<'_>,
+        substream: &mut Self::Substream,
+        buf: &[u8],
+    ) -> Poll<Result<usize, Self::Error>> {
+        use quinn_proto::WriteError;
+        let mut inner = self.inner();
```

Comment:
yamux [has the same problem](https://github.com/paritytech/yamux/blob/d60fca82f1e3a80079a398acabfec1fca561357c/src/connection/stream.rs#L209).  Three fixes come to mind:

* Change `StreamMuxer` methods to take `&mut self`.
* Remove the requirement that `StreamMuxer` instances be `Send + Sync`.
* Make all methods on `StreamMuxer` async so that they can be implemented by a background task and message passing.

Reply:
> yamux has the same problem.

No. Streams do not compete for a connection mutex. The Mutex you linked to protects the stream itself, not the connection.

Reply:
> > yamux has the same problem.
> 
> No. Streams do not compete for a connection mutex. The Mutex you linked to protects the stream itself, not the connection.

Fixing this would require changing `quinn-proto` to use fine-grained locking internally.  `quiche` is worse: its equivalent to `quinn_proto::Endpoint` is `!Send` and `!Sync`, so it can only be used in the thread that created it.  Making this change is desirable but out of scope for this PR.

Reply:
> Fixing this would require changing `quinn-proto` to use fine-grained locking internally.

I do not see how this follows, but ¯\_(ツ)_/¯

Reply:
A `quinn_proto::StreamId` is really just an integer.  Reading or writing from it requires calling methods on `quinn_proto::Connection`.  These methods modify state within the connection, and as such require unique references to the connection object.  `quinn_proto` does no synchronization internally, so I need to provide it externally.  `yamux` does fine-grained synchronization internally, so no external synchronization is needed.

Reply:
(also, you forgot to backslash-escape your `\`)

Reply:
If the `quinn_proto::Connection` is not shared but instead the streams send and receive data over channels the connection does not need to be locked.

Reply:
This is a trade-off: it requires making a full copy in `read_substream` and `write_substream`, and also makes error reporting more difficult.  Furthermore, it requires a significant redesign of the code, so I would prefer to avoid it unless benchmarks show it to be a bottleneck in practice.